zerver/views/tusd.py

```diff
@@ -0,0 +1,201 @@
+        check_upload_within_quota(user_profile.realm, data.size)
+    except RealmUploadQuotaError as e:
+        return reject_upload(str(e))
+
+    # Determine the path_id to store it at
+    file_name = sanitize_name(data.meta_data.get("filename", ""))
+    path_id = upload_backend.generate_message_upload_path(str(user_profile.realm_id), file_name)
+    return tusd_json_response({"ChangeFileInfo": {"ID": path_id}})
+
+
+def handle_upload_pre_finish_hook(
+    request: HttpRequest, user_profile: UserProfile, data: TusUpload
+) -> HttpResponse:
+    metadata = data.meta_data
+    filename = metadata["filename"]
+    # We want to store as the filename a version that clients are
+    # likely to be able to accept via "Save as..."
+    if filename in {"", ".", ".."}:
+        filename = "uploaded-file"
+
+    content_type = metadata.get("filetype")
+    if not content_type:
+        content_type = guess_type(filename)[0]
+        if content_type is None:
+            content_type = "application/octet-stream"
+
+    # With an S3 backend, the filename we passed in pre_create's
+    # data.id has a randomly-generated "mutlipart-id" appended with a
+    # `+`.  Our path_ids cannot contain `+`, so we strip any suffix
+    # starting with `+`.
+    path_id = data.id.partition("+")[0]
+
+    with transaction.atomic():
+        create_attachment(
+            filename,
+            path_id,
+            content_type,
+            attachment_vips_source(path_id),
+            user_profile,
+            user_profile.realm,
+        )
+
+    path = "/user_uploads/" + path_id
+    return tusd_json_response(
+        {
+            "HttpResponse": {
+                "StatusCode": 200,
+                "Body": orjson.dumps({"uri": path, "url": path, "filename": filename}).decode(),
+                "Header": {
+                    "Content-Type": "application/json",
+                },
+            },
+        }
+    )
+
+
+@csrf_exempt
+@typed_endpoint
+def handle_tusd_hook(
+    request: HttpRequest,
+    *,
+    payload: JsonBodyPayload[TusHook],
+) -> HttpResponse:
+    # Make sure this came from localhost
+    if not is_local_addr(request.META["REMOTE_ADDR"]):
+        raise AccessDeniedError
+
+    # Attempt to authenticate via API key or session
+    maybe_user: UserProfile | AnonymousUser = AnonymousUser()
+    if "Authorization" in request.headers:
+        try:
+            role, api_key = get_basic_credentials(request)
+            maybe_user = validate_api_key(
+                request,
+                role,
+                api_key,
+            )
+
+        except JsonableError:
+            pass
+    if isinstance(maybe_user, AnonymousUser):
+        maybe_user = request.user
+
+    hook_name = payload.type
+    if hook_name not in {"pre-create", "pre-finish"}:
+        return reject_upload("Unknown hook")
+
+    maybe_user = request.user
+    if isinstance(maybe_user, AnonymousUser):
+        return reject_upload("Unauthenticated upload")
```

Comment:
The ordering here is a bit odd, to determine/auth the user, then check `payload.type`, then check the user is authenticated. Oh and then `maybe_user = request.user` overrides the above parsing.

I think we could validate `hook_name` either fully before or fully after looking up the user? I guess it could be just a nice `else` clause in the function dispatch below.